File: backend/app/schemas/eval.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class EvaluationCase(BaseModel):
    """Input payload passed to a judge call."""

    prompt: str
    output: str
    reference: str | None = None
    retrieved_context: str | None = None
    conversation_history: list[dict[str, str]] | None = None
    metadata: dict[str, str] = Field(default_factory=dict)

    model_config = {"extra": "forbid"}

    @field_validator("conversation_history")
    @classmethod
    def _validate_conversation_history(
        cls, history: list[dict[str, str]] | None
    ) -> list[dict[str, str]] | None:
        if history is None:
            return None
        for turn in history:
            if set(turn.keys()) != {"role", "content"}:
                raise ValueError(
                    "conversation_history entries must contain exactly 'role' and 'content' keys"
                )
        return history


class Score(BaseModel):
    """Output payload returned by a judge call.

    Field order is load-bearing: `reasoning` must precede `verdict` so that
    OpenAI Structured Outputs / `instructor` emit chain-of-thought reasoning
    before the verdict token (G-Eval / Critique Shadowing, R3 / R7).
    """

    reasoning: str = Field(..., min_length=1)
    verdict: Literal["pass", "fail"]
    per_criterion: dict[str, bool]
    cost_usd: float = Field(default=0.0, ge=0.0)
    latency_ms: int = Field(..., ge=0)
    judge_model: str = Field(..., min_length=1)
    rubric_version: str = Field(..., min_length=1)
    response_hash: str = Field(..., pattern=r"^[a-f0-9]{64}$")

    model_config = {"extra": "forbid"}

    @field_validator("reasoning")
    @classmethod
    def _reasoning_not_blank(cls, reasoning: str) -> str:
        if not reasoning.strip():
            raise ValueError("reasoning must not be empty or whitespace-only")
        return reasoning


class EvalRunCreate(BaseModel):
    """REST body for POST /api/v1/eval/runs."""

    project_id: str = Field(min_length=1)
    training_run_id: str | None = None
    rubric_version_ids: list[str] = Field(min_length=1)
    max_cost_usd: float | None = Field(default=None, gt=0.0)

    model_config = ConfigDict(extra="forbid")


class EvalRunSummary(BaseModel):
    """Row-level eval_run projection used across list and detail responses."""

    id: str
    training_run_id: str | None
    status: str
    started_at: str
    completed_at: str | None
    pass_rate: float | None
    total_cost_usd: float
    max_cost_usd: float | None

    model_config = ConfigDict(from_attributes=True)


class EvalRunListResponse(BaseModel):
    items: list[EvalRunSummary]
    total: int
    limit: int
    offset: int


class EvaluationCasePayload(BaseModel):
    """Structured representation of an eval_case.case_input blob."""

    prompt: str
    output: str
    reference: str | None = None
    retrieved_context: str | None = None
    conversation_history: list[dict[str, str]] | None = None
    metadata: dict[str, str] = Field(default_factory=dict)


class EvalCaseRow(BaseModel):
    """Response shape for an eval_cases row as consumed by the frontend."""

    id: str
    eval_run_id: str
    case_input: EvaluationCasePayload
    input_hash: str


class EvalCallRow(BaseModel):
    """Response shape for an eval_calls row as consumed by the frontend."""

    id: str
    eval_run_id: str
    case_id: str
    rubric_version_id: str
    judge_model: str
    tier: str
    verdict: str
    reasoning: str
    per_criterion: dict[str, bool] | None
    response_hash: str
    cost_usd: float
    latency_ms: int | None
    replayed_from_id: str | None
    created_at: str


class EvalRunDetailResponse(BaseModel):
    run: EvalRunSummary
    cases: list[EvalCaseRow]
    calls: list[EvalCallRow]


class EvalCallsPageResponse(BaseModel):
    items: list[EvalCallRow]
    total: int
    limit: int
    offset: int


class RubricVersionSummary(BaseModel):
    id: str
    rubric_id: str
    version_number: int
    content_hash: str
    calibration_status: str
    judge_model_pin: str
    created_at: str

    model_config = ConfigDict(from_attributes=True)


class RubricSummary(BaseModel):
    id: str
    name: str
    description: str
    research_basis: str | None
    versions: list[RubricVersionSummary]
    created_at: str

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/eval.ts
export type EvalRunStatus = "pending" | "running" | "completed" | "failed" | "cancelled";

export type JudgeVerdict = "pass" | "fail";

export type JudgeTier = "tier1" | "llm";

export type RubricCalibrationStatus = "uncalibrated" | "calibrated" | "failed";

export interface ConversationTurn {
  readonly role: string;
  readonly content: string;
}

export interface EvaluationCasePayload {
  readonly prompt: string;
  readonly output: string;
  readonly reference: string | null;
  readonly retrievedContext: string | null;
  readonly conversationHistory: ReadonlyArray<ConversationTurn> | null;
  readonly metadata: Readonly<Record<string, string>>;
}

export interface EvalRun {
  readonly id: string;
  readonly trainingRunId: string | null;
  readonly status: EvalRunStatus;
  readonly startedAt: string;
  readonly completedAt: string | null;
  readonly passRate: number | null;
  readonly totalCostUsd: number;
  readonly maxCostUsd: number | null;
}

export interface EvalCase {
  readonly id: string;
  readonly evalRunId: string;
  readonly caseInput: EvaluationCasePayload;
  readonly inputHash: string;
}

export interface EvalCall {
  readonly id: string;
  readonly evalRunId: string;
  readonly caseId: string;
  readonly rubricVersionId: string;
  readonly judgeModel: string;
  readonly tier: JudgeTier;
  readonly verdict: JudgeVerdict;
  readonly reasoning: string;
  readonly perCriterion: Readonly<Record<string, boolean>> | null;
  readonly responseHash: string;
  readonly costUsd: number;
  readonly latencyMs: number | null;
  readonly replayedFromId: string | null;
  readonly createdAt: string;
}

export interface EvalRunDetail {
  readonly run: EvalRun;
  readonly cases: ReadonlyArray<EvalCase>;
  readonly calls: ReadonlyArray<EvalCall>;
}

export interface EvalCallsPage {
  readonly items: ReadonlyArray<EvalCall>;
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
}

export interface RubricVersion {
  readonly id: string;
  readonly rubricId: string;
  readonly versionNumber: number;
  readonly contentHash: string;
  readonly calibrationStatus: RubricCalibrationStatus;
  readonly judgeModelPin: string;
  readonly createdAt: string;
}

export interface Rubric {
  readonly id: string;
  readonly name: string;
  readonly description: string;
  readonly researchBasis: string | null;
  readonly versions: ReadonlyArray<RubricVersion>;
  readonly createdAt: string;
}

export interface EvalRunCreateRequest {
  readonly trainingRunId: string | null;
  readonly rubricVersionIds: ReadonlyArray<string>;
  readonly maxCostUsd: number | null;
}

export interface EvalListQuery {
  readonly trainingRunId?: string | null;
  readonly limit?: number;
  readonly offset?: number;
}

export interface EvalCaseCompletedPayload {
  readonly evalRunId: string;
  readonly caseId: string;
  readonly rubricVersionId: string;
  readonly evalCallId: string;
  readonly verdict: JudgeVerdict;
  readonly costUsd: number;
  readonly latencyMs: number | null;
}

export interface EvalRunCompletedPayload {
  readonly evalRunId: string;
  readonly status: EvalRunStatus;
  readonly passRate: number | null;
  readonly totalCostUsd: number;
}

export interface EvalCostWarningPayload {
  readonly evalRunId: string;
  readonly currentCostUsd: number;
  readonly maxCostUsd: number;
  readonly warningPct: number;
}

export type EvalEventName = "case_completed" | "run_completed" | "cost_warning";
